ibis/backends/__init__.py

```diff
@@ -777,7 +778,76 @@ def drop_schema(
         self.drop_database(name=name, catalog=database, force=force)
 
 
-class BaseBackend(abc.ABC, _FileIOHandler):
+class _CacheEntry(NamedTuple):
+    name: str
+    orig_op: ops.Relation
+    cached_op_ref: weakref.ref[ops.Relation]
+    finalizer: weakref.finalize
+
+
+class _CacheHandler:
+    """A mixin for handling `.cache()`/`CachedTable` operations."""
+
+    def __init__(self):
+        self._cache_name_to_entry = {}
+        self._cache_op_to_entry = {}
+
+    def _cached_table(self, table: ir.Table) -> ir.CachedTable:
+        """Convert a Table to a CachedTable.
+
+        Parameters
+        ----------
+        table
+            Table expression to cache
+
+        Returns
+        -------
+        Table
+            Cached table
+        """
+        entry = self._cache_op_to_entry.get(table.op())
+        if entry is None or (cached_op := entry.cached_op_ref()) is None:
+            cached_op = self._create_cached_table(util.gen_name("cached"), table).op()
+            entry = _CacheEntry(
+                cached_op.name,
+                table.op(),
+                weakref.ref(cached_op),
+                weakref.finalize(
+                    cached_op, self._finalize_cached_table, cached_op.name
+                ),
+            )
+            self._cache_op_to_entry[table.op()] = entry
+            self._cache_name_to_entry[cached_op.name] = entry
+        return ir.CachedTable(cached_op)
+
+    def _finalize_cached_table(self, name: str) -> None:
+        """Release a cached table given its name.
+
+        This is a no-op if the cached table is already released.
+
+        Parameters
+        ----------
+        name
+            The name of the cached table.
+        """
+        if (entry := self._cache_name_to_entry.pop(name, None)) is not None:
```

Comment:
Isn't `name` equivalent to `entry.name`? If so, can't we get rid of storing `name` as a field on `_CacheEntry`?

Reply:
Nice catch, fixed.